contracts/staking/ServiceStakingBase.sol

```diff
@@ -0,0 +1,278 @@
+                ServiceInfo storage curInfo = mapServiceInfo[curServiceId];
+
+                // Calculate the staking nonce ratio
+                uint256 curNonce = IMultisig(curInfo.multisig).nonce();
+                // Calculate the nonce ratio in 1e18 value
+                uint256 ratio = ((block.timestamp - curInfo.ts) * 1e18) / (curNonce - curInfo.nonce);
+
+                // Record the reward for the service if it has provided enough transactions
+                if (ratio >= stakingRatio) {
+                    eligibleServiceIds[numServices] = curServiceId;
+                    ++numServices;
+                } else {
+                    // Record current timestamp for each reward ineligible service
+                    curInfo.ts = block.timestamp;
+                }
+                // Record current service multisig nonce
+                curInfo.nonce = curNonce;
+
+                // Record the unstaked service Id index in the global set of staked service Ids
+                if (curServiceId == serviceId) {
+                    idx = i;
+                }
+            }
+
+            // Calculate each eligible service Id reward
+            uint256 tsLast = tsLastBalance;
+            // Calculate the maximum possible reward per service during the last deposit period
+            uint256 maxRewardsPerService = (rewardsPerSecond * (block.timestamp - tsLast)) / numServices;
+            // Traverse all the eligible services and calculate their rewards
+            for (uint256 i = 0; i < numServices; ++i) {
+                uint256 curServiceId = eligibleServiceIds[i];
+                ServiceInfo storage curInfo = mapServiceInfo[curServiceId];
+
+                // Calculate the reward up until now
+                // If the staking was longer than the deposited period, the service's timestamp is adjusted such that
+                // it is equal to at most the tsLastBalance of the last deposit happening during every _checkpoint() call
+                uint256 reward = rewardsPerSecond * (block.timestamp - curInfo.ts);
+                // Adjust the reward if it goes out of calculated max bounds
+                if (reward > maxRewardsPerService) {
+                    reward = maxRewardsPerService;
+                }
+
+                // Adjust the deposit balance
+                if (lastBalance >= reward) {
+                    lastBalance -= reward;
+                } else {
+                    // This situation must never happen
+                    // TODO: Fuzz this
+                    reward = lastBalance;
+                    lastBalance = 0;
+                }
+
+                // Add the calculated reward to the service info
+                curInfo.reward += reward;
+
+                // Adjust the starting ts for each service to a current timestamp
+                curInfo.ts = block.timestamp;
+            }
+            // Update the storage balance
+            balance = lastBalance;
```

Comment:
We should probably have two balance variables: `balance` should simply show the total balance currently in the contract. Then `availableRewards` is the balance that's still available to stakers. Again, good to have this separate for frontends and to easily inspect